tests/installation/select_patterns_and_packages.pm

```diff
@@ -155,113 +141,127 @@ sub package_action {
+    delete $patterns{all};
+    while (1) {
+        die "looping for too long" unless ($counter--);
+        my $needs_to_be_selected;
+        my $ret = check_screen('on-pattern', 1);
+        # this variable will only be updated when the pattern list entry is under the cursor
+        # and the pattern is in the PATTERNS variable and there is a needle for that entry
+        my $current_pattern = 'UNKNOWN_PATTERN';
 
-            if (get_var('PATTERNS', '') =~ /default/ && !(get_var('PATTERNS', '') =~ /$current_pattern/)) {
-                $needs_to_be_selected = $selected;
-                record_info("keep default", "");
-            }
-            if (!$needs_to_be_selected && $selected) {
-                send_key ' ';
-                record_info("unselect", "");
-                assert_screen [qw(current-pattern-unselected current-pattern-autoselected)], 8;
+        if ($ret) {    # unneedled pattern
+            for my $p (keys %patterns) {
+                my $sel = 1;
+                if ($p =~ /^-/) {
+                    # this pattern shall be deselected as indicated by '-' prefix
+                    $sel = 0;
+                    $p =~ s/^-//;
+                }
+                if (match_has_tag("pattern-$p")) {
+                    $needs_to_be_selected = $sel;
+                    $current_pattern      = $p;
+                    # add pattern to the set of detected patterns
+                    $processed_patterns{$p} = undef;
+                    delete $patterns{$p};
+                    record_info($current_pattern, $needs_to_be_selected);
+                }
             }
-            movedownelseend;
-            check12qtbug;
+            last if ((get_var('PATTERNS', '') =~ /default/) && !(scalar keys %patterns));
+        }
+
+        $needs_to_be_selected = 1 if get_var('PATTERNS', '') =~ /all/;
+
+        my $selected = check_screen([qw(current-pattern-selected on-category)], 0);
+        if ($selected && $selected->{needle}->has_tag('on-category')) {
+            move_down;
+            next;
+        }
+        if ($needs_to_be_selected && !$selected) {
+            switch_selection('select');
+            assert_screen 'current-pattern-selected', 5;
         }
-        # check if we have processed all patterns mentioned in the test suite settings
-        # delete special 'all' and 'default' keys from the check
-        delete $patterns{default};
-        delete $patterns{all};
-        my @unseen;
-        foreach my $k (keys %patterns) {
-            #Remove - from the key in case of deselection
-            $k =~ s/^-//;
-            push @unseen, $k if (not exists $processed_patterns{$k});
+        if (get_var('PATTERNS', '') =~ /default/ && !(get_var('PATTERNS', '') =~ /$current_pattern/)) {
+            $needs_to_be_selected = $selected;
+            record_info("keep default", "");
         }
-        die "Not all patterns given in the job settings were processed:" . join(", ", @unseen) if @unseen;
```

Comment:
You have removed that check completely, which is bad. We could reuse you logic here that we simply check if any keys are remaining in the hash though, that would be a way to go.

Reply:
Added back.